src/main/java/mate/academy/bookstore/service/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,95 @@
+package mate.academy.bookstore.service;
+
+import lombok.RequiredArgsConstructor;
+import mate.academy.bookstore.dto.cartitem.CreateCartItemRequestDto;
+import mate.academy.bookstore.dto.cartitem.UpdateCartItemRequestDto;
+import mate.academy.bookstore.dto.shoppingcart.ShoppingCartDto;
+import mate.academy.bookstore.exception.EntityNotFoundException;
+import mate.academy.bookstore.mapper.ShoppingCartMapper;
+import mate.academy.bookstore.model.Book;
+import mate.academy.bookstore.model.CartItem;
+import mate.academy.bookstore.model.ShoppingCart;
+import mate.academy.bookstore.model.User;
+import mate.academy.bookstore.repository.book.BookRepository;
+import mate.academy.bookstore.repository.cartitem.CartItemRepository;
+import mate.academy.bookstore.repository.shoppingcart.ShoppingCartRepository;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final CartItemRepository cartItemRepository;
+    private final BookRepository bookRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+
+    @Override
+    @Transactional
+    public ShoppingCartDto getShoppingCartForUser(Long userId) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId)
+                .orElseThrow(() -> new EntityNotFoundException("Shopping cart not found "
+                        + "for user id " + userId));
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    @Transactional
+    public ShoppingCartDto addItemToCart(CreateCartItemRequestDto requestDto, Long userId) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId)
+                .orElseThrow(() -> new EntityNotFoundException("Shopping cart not found "
+                        + "for user id " + userId));
+
+        Book book = bookRepository.findById(requestDto.getBookId())
+                .orElseThrow(() -> new EntityNotFoundException("Book not found for id "
+                        + requestDto.getBookId()));
+
+        CartItem cartItem = cartItemRepository.findByShoppingCartAndBook(shoppingCart, book)
+                .orElse(null);
```

Comment:
orElseThrow?

Reply:
we dont use orElseThrow because the logic is that if cartItem is null, it means we need to create it

Reply:
in this case you can use orElseGet and provide a supplier